src/analyze/analyze-shutdown-blame.c

```diff
@@ -0,0 +1,227 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "sd-bus.h"
+#include "sd-journal.h"
+
+#include "alloc-util.h"
+#include "analyze.h"
+#include "analyze-shutdown-blame.h"
+#include "bus-util.h"
+#include "format-table.h"
+#include "hashmap.h"
+#include "journal-util.h"
+#include "log.h"
+#include "runtime-scope.h"
+#include "sort-util.h"
+#include "string-util.h"
+#include "time-util.h"
+
+typedef struct ShutdownTime {
+        char *name;
+        usec_t start_time;
+        usec_t stop_time;
+        usec_t duration;
+} ShutdownTime;
+
+static ShutdownTime* shutdown_time_free(ShutdownTime *t) {
+        if (!t)
+                return NULL;
+
+        free(t->name);
+        return mfree(t);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(ShutdownTime*, shutdown_time_free);
+
+static int compare_shutdown_times(ShutdownTime * const *a, ShutdownTime * const *b) {
+        if ((*a)->duration > (*b)->duration)
+                return -1;
+        if ((*a)->duration < (*b)->duration)
+                return 1;
+        return 0;
+}
+
+static int acquire_shutdown_times(Hashmap **ret) {
+        _cleanup_(sd_journal_closep) sd_journal *j = NULL;
+        _cleanup_(hashmap_freep) Hashmap *shutdown_times = NULL;
+        const void *data;
+        size_t length;
+        int r;
+
+        shutdown_times = hashmap_new(&string_hash_ops);
+        if (!shutdown_times)
+                return log_oom();
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open journal: %m");
+
+        /* Look for systemd messages with units */
+        r = sd_journal_add_match(j, "SYSLOG_IDENTIFIER=systemd", 0);
```

Comment:
hmm SYSLOG_IDENTIFIER is not a suitable match here. it's kinda a legacy field, and user controlled, not journald controlled...